runtime/smithy-client/common/src/aws/smithy/kotlin/runtime/client/AbstractSdkClientFactory.kt

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.client
+
+public abstract class AbstractSdkClientFactory<
+    TConfig : SdkClientConfig,
+    TConfigBuilder : SdkClientConfig.Builder<TConfig>,
+    TClient : SdkClient,
+    TClientBuilder : SdkClient.Builder<TConfig, TConfigBuilder, TClient>,
+    > : SdkClientFactory<TConfig, TConfigBuilder, TClient, TClientBuilder> {
+
+    /**
+     * Inject any client-specific config
+     */
+    protected open fun finalizeConfig(builder: TClientBuilder) { }
+
+    public override operator fun invoke(block: TConfigBuilder.() -> Unit): TClient = builder().apply {
+        config.apply(block)
+        finalizeConfig(this)
+    }.build()
```

Comment:
Invoking an abstract class looks a little odd since they aren't meant to be instantiated. I don't have any suggestion here, it just stands out to me

Reply:
This isn't quite invoking an abstract class. It's an operator defined on an abstract class which will only be callable on a concrete instance. No one's ever going to call `AbstractSdkClientFactory.invoke { … }` or `AbstractSdkClientFactory().invoke { … }`—both are compile-time errors.

This simply enables classes which inherit `AbstractSdkClientFactory` (namely, service client companion objects) to be invoked (e.g., `S3Client.invoke { … }`).